lib/loginator/transaction.rb

```diff
@@ -1,21 +1,125 @@
 require 'securerandom'
 
 module Loginator
-  # Methods for generating transactional metadata.
-  module Transaction
-    def self.from_json(json_str)
-      json = MultiJson.load(json_str)
-      Loginator.const_get(json['type'].capitalize).from_hash(json)
+  # A Loginator::Transaction is meant to encapsulate a single
+  # request/response cycle.
+  class Transaction
+    class << self
+      # @param [String] json JSON body of the given transaction
+      # @return [Transaction]
+      def from_json(json)
+        obj = MultiJson.load(json, symbolize_keys: true)
+        Transaction.new(filter_hash(obj))
+      end
+
+      protected
+
+      # This is the inverse functions for the instance-level
+      # filter_hash.
+      def filter_hash(hsh)
+        [[:timestamp, ->(t) { Time.at(t) }]].map do |k, f|
+          hsh[k] = f.call(hsh[k])
+        end
+        hsh
+      end
+    end
+
+    attr_accessor :path, :request, :params, :response, :status
+
+    attr_reader :uuid, :timestamp, :duration
+
+    protected
+
+    attr_writer :uuid, :timestamp, :duration
+
+    public
+
+    # Create a new Loginator::Transaction
+    # @param [Hash] opts The options to create the Loginator::Transaction
+    # @option opts [String]  :uuid UUID (SecureRandom.uuid)
+    # @option opts [Time]    :timestamp (Time.now) Beginning timestamp of transaction
+    # @option opts [Integer] :status Status returned to the requester
+    # @option opts [String]  :path  Path associated with the request
+    # @option opts [String]  :request Body of the request
+    # @option opts [Hash]    :params Parameters of the request
+    # @option opts [String]  :response Body of the response
+    # @option opts [Float]   :duration Duration of the request
+    def initialize(opts = {})
+      @uuid = opts.delete(:uuid) || SecureRandom.uuid
+      @timestamp = opts.delete(:timestamp) || Time.now
+      @duration = opts.delete(:duration)
+      opts.each_pair do |k, v|
+        send("#{k}=", v)
+      end
+    end
+
+    # Marks the beginning of a Transaction. Optionally, will execute
+    # the given block in the context of a transaction, returning the
+    # last line of the block and raising any exceptions thrown in the
+    # block given.
+    #
+    # NOTE: we make a best effort to guarantee that `transaction.finished`
+    # is called via an ensure block, but that is all. We do not set the
+    # status or response. You must do so in your block if you wish to
+    # record any failures.
+    # @param [Block] &blk optional
+    def begin
+      @timestamp = Time.now
+      yield self if block_given?
```

Comment:
*sigh* So this smells to me, but I'm okay with it, because the block is evaluated with the binding of the caller and not of this method, so you don't break the fourth wall.